tortoise/algorithm.go

```diff
@@ -147,6 +146,20 @@ func (t *Tortoise) LatestComplete() types.LayerID {
 	return t.trtl.verified
 }
 
+func (t *Tortoise) Updates() (types.LayerID, []types.BlockContextualValidity) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	res := make([]types.BlockContextualValidity, len(t.trtl.updated))
+	copy(res, t.trtl.updated)
+	return t.trtl.verified, res
+}
+
+func (t *Tortoise) UpdatesPersisted(persisted []types.BlockContextualValidity) {
```

Comment:
i think this interface is not necessary. tortoise created an update event, and it won't create it again unless opinion changes. it doesn't need to know if that event was persisted

if you want persisted can be renamed to emitted

Reply:
ok. but there is one corner case:

when mesh process the updates, and it fails to revert the state accordingly, those block validity will not be persisted to database. in that case, tortoise also won't emit those updates again. in that case, mesh will not try to check those layers again, until syncer realizes it has different mesh hash.

i could accumulate the updates in mesh.go if the failure happen and remove this interface.